src/app/clusters/bindings/BindingManager.h

```diff
@@ -0,0 +1,195 @@
+
+        void AddPendingNotification(EndpointId endpoint, ClusterId cluster, void * context)
+        {
+            for (ClusterPath & path : *this)
+            {
+                // New notifications for the same (endpoint, cluster) shall
+                // simply overrride the old ones
+                if (path.cluster == cluster && path.endpoint == endpoint)
+                {
+                    path.context = context;
+                    return;
+                }
+            }
+            if (mNumPendingNotifications < kMaxPendingNotifications)
+            {
+                mPendingNotifications[mNumPendingNotifications++] = { endpoint, cluster, context };
+            }
+            else
+            {
+                mPendingNotifications[mNextToOverride] = { endpoint, cluster, context };
+                mNextToOverride++;
+                mNextToOverride %= kMaxPendingNotifications;
+            }
+        }
+
+    private:
+        uint8_t mNumPendingNotifications = 0;
+        uint8_t mNextToOverride          = 0;
+
+        PeerId mPeerId;
+        ClusterPath mPendingNotifications[kMaxPendingNotifications];
+        System::Clock::Timestamp mLastUpdateTime;
+    };
+
+    // The pool for all the pending comands.
+    class PendingNotificationMap
+    {
+    public:
+        PendingNotificationEntry * FindLRUEntry();
+
+        PendingNotificationEntry * FindEntry(PeerId peerId);
+
+        CHIP_ERROR AddPendingNotification(PeerId peer, EndpointId endpoint, ClusterId cluster, void * context);
+
+        void RemoveEntry(PendingNotificationEntry * entry) { mPendingNotificationMap.ReleaseObject(entry); }
+
+        template <typename Function>
+        Loop ForEachActiveObject(Function && function)
+        {
+            return mPendingNotificationMap.ForEachActiveObject(std::forward<Function>(function));
+        }
+
+    private:
+        BitMapObjectPool<PendingNotificationEntry, EMBER_BINDING_TABLE_SIZE> mPendingNotificationMap;
+    };
+
+    static void HandleDeviceConnected(void * context, OperationalDeviceProxy * device);
+    void HandleDeviceConnected(OperationalDeviceProxy * device);
+
+    static void HandleDeviceConnectionFailure(void * context, PeerId peerId, CHIP_ERROR error);
+    void HandleDeviceConnectionFailure(PeerId peerId, CHIP_ERROR error);
+
+    CHIP_ERROR EstablishConnection(FabricIndex fabric, NodeId node);
+
+    // Called when CASE session is established to a peer device. Will send all the pending commands to the peer.
+    void SyncPendingNotificationsToPeer(OperationalDeviceProxy * device, PendingNotificationEntry * pendingClusters);
+
+    // Called when CASE session is not established to a peer device. Will enqueue the command and initialize connection.
+    CHIP_ERROR EnqueueUnicastNotification(FabricIndex fabric, NodeId node, EndpointId endpoint, ClusterId cluster, void * context);
+
+    PendingNotificationMap mPendingNotificationMap;
```

Comment:
One other thing.  This PR is aadding >1KB of .bss use.  That seems like quite a lot.  Please file a followup to figure out the breakdown of that 1KB and reduce it, because I don't think we can ship like this; we're too close to our RAM use boundaries as it is.

Reply:
The main .bss increase is occupied by the `mPendingNotificationMap` object. It contains typically 10 entried each with a queue for pending notifications to be sent to peer. The queue size is reduced from 5 to 3. Each entry is 80 bytes now and the whole map occupies 856 bytes. In the uploaded patch, the `BindingManager` instance is now set by the application. 

For devices without binding feature such as simple LED lights, the size increase is eliminated. For devices with binding I believe the size increase is kind of a must.

Reply:
So I assume the size of a PendingNotificationEntry (on a 32-bit system that can't 4-byte-align a uint64_t, so arm32) comes from:
* 8 bytes for the two uint8_t (due to padding after them), I assume.
* 16 bytes PeerId (8 bytes NodeId, 8 bytes CompressedFabricId).
* 36 bytes mPendingNotifications (ClusterPath is 12 bytes due to padding after the endpoint id)
* 4 bytes padding.
* 8 bytes timestamp.

That adds up to 72 bytes.  That's what I measure on esp32 with static_assert.

There are a few things we can do to reduce the size of this:

1. Most trivially, reordering so the order is "PeerId, Timestamp, mPendingNotifications, uint8_t, uint8_t" will get rid of some of the padding and drop the size to 64 bytes.
2. Slightly more work: Instead of storing a PeerId we could store a NodeId and FabricIndex.  If we put the FabricIndex next to the two uint8_t, that will drop us to 56 bytes.
3. Slightly more work: just store a list of (index into binding table, void *) pairs, with the restriction that any given PendingNotificationEntry stores only indices that point to the same peer.  This requires a notification when entries get removed, so we can clear out the corresponding list elements.  But from a binding entry index we can retrieve the PeerId (from the fabric index and node id of the entry), the local endpoint and cluster id (so everything in the ClusterPath except the void*).  If we do that, we will have 8 bytes for timestamp, 24 bytes for the three list entries, our two uint8_t, plus padding, for a total of 40 bytes.
4. Slightly more work: do the above, but have two separate lists of 3 indexes and 3 void*.  If we order these as (timestamp, list of void*, list of indices, uint8_t, uint8_t), then we have 8 + 3 * 4 + 3 + 2 + padding == 28 bytes.

If we do _that_, then we might also want to just get rid of the grouping by peer and have a single flat list of some capped size, so we don't need to worry about whether we have lots of changes queued for one peer, or one change queued for a bunch of peers....  That might save some space too.

All of this is followup fodder, but I am quite confident that we can do better than what we have right now.

As far as the change to have the BindingManager be set by the app, I'd think we want it created automatically, as before, but just conditioned on whether support for the bindings cluster is compiled in.

Reply:
The trivial struct reordering is added in-place and further improvement request submitted at https://github.com/project-chip/connectedhomeip/issues/13749.